### PR TITLE
Feature/hw 7

### PR DESCRIPTION
**Pull request by rosklyar**

### CODE REVIEW

File: .github/workflows/labelling-bot-deploy.yaml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+              sudo yum update -y
+              sudo yum install docker -y
+              sudo service docker start
+              sudo usermod -a -G docker ec2-user
+            fi
+            
+            # Start Docker service if not running
+            if ! sudo service docker status &> /dev/null; then
+              sudo service docker start
+            fi
```
Comment (truskovskiyk) [inline .github/workflows/labelling-bot-deploy.yaml:51]: why do you need to do this every run? what is the reason to use our own VM for this?

File: garbage_classifier/garbage_classifier/garbage_data.py [hunk 2]
```diff
@@ -34,22 +38,112 @@ def __len__(self):
...
 
-    def get_val_loader(self):
-        return self._val_loader
\ No newline at end of file
+    def get_test_loader(self):
+        return self._test_loader
+    
+def load_train_data(s3_access_key, s3_secret_key, s3_bucket, s3_prefix):
+    s3_client = boto3.client('s3', aws_access_key_id=s3_access_key, aws_secret_access_key=s3_secret_key)
+    folder = 'data/real-ds'
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/garbage_data.py]: remove hardcode

File: garbage_classifier/garbage_classifier/garbage_data.py [hunk 3]
```diff
@@ -34,22 +38,112 @@ def __len__(self):
...
+    _download_files_from_s3_bucket_folder(s3_client, s3_bucket, s3_prefix, f'{folder}/input')
+    _create_train_test_split(f'{folder}/input', f'{folder}/train', f'{folder}/test')
+    _create_tar_gz_folder(f'{folder}/train', 'data/real-ds/train.tar.gz')
+    _create_tar_gz_folder(f'{folder}/test', 'data/real-ds/test.tar.gz')
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/garbage_data.py]: please remove hardcode, give names

File: garbage_classifier/garbage_classifier/garbage_data.py [hunk 4]
```diff
@@ -34,22 +38,112 @@ def __len__(self):
...
+
+def load_data(s3_access_key, s3_secret_key, s3_bucket, s3_prefix):
+    s3_client = boto3.client('s3', aws_access_key_id=s3_access_key, aws_secret_access_key=s3_secret_key)
+    folder = 'data/tmp'
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/garbage_data.py]: same

File: garbage_classifier/garbage_classifier/garbage_data.py [hunk 5]
```diff
@@ -34,22 +38,112 @@ def __len__(self):
...
+
+        images = [f for f in os.listdir(src_class_folder) if os.path.isfile(os.path.join(src_class_folder, f))]
+        np.random.shuffle(images)
+        split_index = int(len(images) * 0.8)
+
+        train_images = images[:split_index]
+        test_images = images[split_index:]
+
+        for img in train_images:
+            shutil.copy(os.path.join(src_class_folder, img), os.path.join(train_class_folder, img))
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/garbage_data.py:94]: consider using [pathlib](https://docs.python.org/3/library/pathlib.html#module-pathlib) instead os.path

File: garbage_classifier/garbage_classifier/garbage_data.py [hunk 6]
```diff
@@ -34,22 +38,112 @@ def __len__(self):
...
+
+                os.makedirs(os.path.dirname(local_file_path), exist_ok=True)
+                print(f'Downloading {file_key} to {local_file_path}')
+                s3.download_file(bucket_name, file_key, local_file_path)
+                print(f'Downloaded {file_key} to {local_file_path}')
+
+def extract_tar_gz(archive_path):
+    if not os.path.exists(archive_path):
+        print(f"The file '{archive_path}' does not exist.")
+        return None
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/garbage_data.py]: don't need to return None, python does this by default

File: garbage_classifier/garbage_classifier/garbage_data.py [hunk 7]
```diff
@@ -34,22 +38,112 @@ def __len__(self):
...
+    if not os.path.exists(extracted_path):
+        os.makedirs(extracted_path)
+
+    try:
+        with tarfile.open(archive_path, 'r:gz') as tar:
+            tar.extractall(path=extracted_path)
+            print(
+                f"Successfully extracted '{archive_path}' to '{extracted_path}'.")
+            return extracted_path
+    except Exception as e:
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/garbage_data.py:160]: please don't use Exception type, specify type of error more clearly

File: garbage_classifier/garbage_classifier/garbage_data.py [hunk 8]
```diff
@@ -34,22 +38,112 @@ def __len__(self):
...
+                print(f'Downloaded {file_key} to {local_file_path}')
+
+def extract_tar_gz(archive_path):
+    if not os.path.exists(archive_path):
+        print(f"The file '{archive_path}' does not exist.")
+        return None
+
+    if not tarfile.is_tarfile(archive_path):
+        print(f"The file '{archive_path}' is not a valid tar archive.")
+        return None
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/garbage_data.py]: just `return`

File: garbage_classifier/garbage_classifier/garbage_data.py [hunk 9]
```diff
@@ -34,22 +38,112 @@ def __len__(self):
...
+
+    try:
+        with tarfile.open(archive_path, 'r:gz') as tar:
+            tar.extractall(path=extracted_path)
+            print(
+                f"Successfully extracted '{archive_path}' to '{extracted_path}'.")
+            return extracted_path
+    except Exception as e:
+        print(f"Error while extracting '{archive_path}': {e}")
+        return None
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/garbage_data.py]: just `return`

File: garbage_classifier/garbage_classifier/inference.py [hunk 10]
```diff
@@ -50,19 +52,32 @@ def multi_process_prediction(model, feature_extractor, dataset: ImageFolder, pro
...
 
+def make_inference(model_path, data_archive_path):
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    model = load_model(model_path, device)
+    feature_extractor = FeatureExtractor("microsoft/beit-base-patch16-224-pt22k-ft22k")
+    data_path = extract_tar_gz(data_archive_path)
+    dataset = ImageFolder(data_path)
+    paths = [path for path, _ in dataset.imgs]
+    multi_process_result = single_process_prediction(model, feature_extractor, dataset, device)
+    with open("result.csv", "w") as f:
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/inference.py:63]: remove hardcode

File: garbage_classifier/garbage_classifier/inference.py [hunk 11]
```diff
@@ -50,19 +52,32 @@ def multi_process_prediction(model, feature_extractor, dataset: ImageFolder, pro
...
+        for path, label in zip(paths, multi_process_result):
+            f.write(f"{path},{label}\n") 
+
 if __name__ == "__main__":
-    model = load_model()
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    model = load_model("model/model.pth", device)
```
Comment (truskovskiyk) [inline garbage_classifier/garbage_classifier/inference.py:69]: remove hardcode please

File: garbage_classifier/pipelines/garbage-classifier-inference-pipeline.yaml [hunk 12]
```diff
@@ -0,0 +1,116 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (truskovskiyk) [inline garbage_classifier/pipelines/garbage-classifier-inference-pipeline.yaml]: remove this yaml please

File: garbage_classifier/pipelines/garbage-classifier-training-pipeline.yaml [hunk 13]
```diff
@@ -0,0 +1,123 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (truskovskiyk) [inline garbage_classifier/pipelines/garbage-classifier-training-pipeline.yaml]: remove yaml please

File: garbage_classifier/pipelines/kubeflow-pipelines.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+def garbage_classifier_traininig_pipeline(s3_access_key, s3_secret_key, s3_bucket, s3_prefix, wandb_api_key):
+
+    load_data = dsl.ContainerOp(
+        name="load_data",
+        command=["python", "garbage_classifier/cli.py", "load-train-data",
+                 s3_access_key, s3_secret_key, s3_bucket, s3_prefix],
+        image=IMAGE,
+        file_outputs={"train": "data/real-ds/train.tar.gz",
+                      "test": "data/real-ds/test.tar.gz"},
+    )
```
Comment (truskovskiyk) [inline garbage_classifier/pipelines/kubeflow-pipelines.py]: try to format this code, consider to use https://github.com/psf/black

File: garbage_classifier/requirements.txt [hunk 15]
```diff
@@ -1,3 +1,4 @@
+boto3
```
Comment (truskovskiyk) [inline garbage_classifier/requirements.txt]: add versions

File: garbage_classifier/tests/test_garbage_data.py [hunk 16]
```diff
@@ -16,16 +16,12 @@ def test_feature_extractor(input_size):
 
 def test_garbage_data():
     batch_size = 2
-    dataset_length = 10
-    val_split = 0.2
     garbage_data = gd.GarbageData(
-        "garbage_classifier/tests/data", batch_size, val_split)
+        "garbage_classifier/tests/data", "garbage_classifier/tests/data", batch_size)
```
Comment (truskovskiyk) [inline garbage_classifier/tests/test_garbage_data.py:20]: remove hardcode please

Comment (truskovskiyk) [reply garbage_classifier/tests/test_garbage_data.py:20]: oo, this is test. okay for tests it's okay to hardcode btw

File: labelling_bot/labelling_bot/bot.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+logging.basicConfig(
+    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO
+)
+logger = logging.getLogger(__name__)
+
+s3 = boto3.client('s3')
+
+# upload config.json to s3
+s3.upload_file('config.json', BUCKET_NAME, f'{ROOT_FOLDER}/config.json')
```
Comment (truskovskiyk) [inline labelling_bot/labelling_bot/bot.py:19]: please avoid global variables

**Review by truskovskiyk — COMMENTED** [review]
@rosklyar , amazing work, please see feedback

File: .github/workflows/labelling-bot-deploy.yaml (hunk 1 above)
Comment (rosklyar) [reply .github/workflows/labelling-bot-deploy.yaml:51]: Just checking if docker is ready on target instance as it was not ready during first ssh deployment :)

File: garbage_classifier/garbage_classifier/garbage_data.py (hunk 2 above)
Comment (rosklyar) [reply garbage_classifier/garbage_classifier/garbage_data.py]: Hardcode removed in latest pr for Airflow https://github.com/rosklyar/projector-ml-ops/pull/13/files#diff-602516c0892d0bcf7a23fc1348d4275fdf4ef1d10e4e476ffb7dba3f39990d1dR61

File: garbage_classifier/garbage_classifier/garbage_data.py (hunk 3 above)
Comment (rosklyar) [reply garbage_classifier/garbage_classifier/garbage_data.py]: Hardcode removed in latest pr for Airflow https://github.com/rosklyar/projector-ml-ops/pull/13/files#diff-602516c0892d0bcf7a23fc1348d4275fdf4ef1d10e4e476ffb7dba3f39990d1dR61

File: garbage_classifier/garbage_classifier/garbage_data.py (hunk 4 above)
Comment (rosklyar) [reply garbage_classifier/garbage_classifier/garbage_data.py]: https://github.com/rosklyar/projector-ml-ops/pull/13/files#diff-602516c0892d0bcf7a23fc1348d4275fdf4ef1d10e4e476ffb7dba3f39990d1dR70

File: garbage_classifier/garbage_classifier/garbage_data.py (hunk 6 above)
Comment (rosklyar) [reply garbage_classifier/garbage_classifier/garbage_data.py]: I think GPT did it to terminate execution :)

File: garbage_classifier/garbage_classifier/garbage_data.py (hunk 8 above)
Comment (rosklyar) [reply garbage_classifier/garbage_classifier/garbage_data.py]: ok

File: garbage_classifier/garbage_classifier/garbage_data.py (hunk 9 above)
Comment (rosklyar) [reply garbage_classifier/garbage_classifier/garbage_data.py]: ok

File: garbage_classifier/garbage_classifier/garbage_data.py (hunk 6 above)
Comment (rosklyar) [reply garbage_classifier/garbage_classifier/garbage_data.py]: Ah, I see now

File: garbage_classifier/pipelines/garbage-classifier-training-pipeline.yaml (hunk 13 above)
Comment (rosklyar) [reply garbage_classifier/pipelines/garbage-classifier-training-pipeline.yaml]: Ok

File: garbage_classifier/pipelines/garbage-classifier-inference-pipeline.yaml (hunk 12 above)
Comment (rosklyar) [reply garbage_classifier/pipelines/garbage-classifier-inference-pipeline.yaml]: Ok

File: garbage_classifier/pipelines/kubeflow-pipelines.py (hunk 14 above)
Comment (rosklyar) [reply garbage_classifier/pipelines/kubeflow-pipelines.py]: ok

File: garbage_classifier/requirements.txt (hunk 15 above)
Comment (rosklyar) [reply garbage_classifier/requirements.txt]: Ok

**Review by truskovskiyk — APPROVED** [review]
@rosklyar , kubeflow part looks good

**Pull request merged**